### PR TITLE
[proxy] Handle null HostConfigs gracefully in start

### LINKED ISSUE #1481: Panic in weaveproxy when starting container with 'null' HostConfig
The weave proxy (master) panics when running the ECS demo. It works with Weave 1.1. 

Here are the weaveproxy logs:

```
INFO: 2015/09/30 11:12:49.620486 weave proxy git-d5c85a3d53f4
INFO: 2015/09/30 11:12:49.620619 Command line arguments: --hostname-from-label com.amazonaws.ecs.container-name -H unix:///var/run/weave.sock
INFO: 2015/09/30 11:12:49.621511 [docker] Using Docker API on unix:///var/run/docker.sock: &[KernelVersion=3.14.48-33.39.amzn1.x86_64 Os=linux Version=1.6.2 ApiVersion=1.18 Arch=amd64 GitCommit=7c8fca2/1.6.2 GoVersion=go1.3.3]
INFO: 2015/09/30 11:12:49.669784 proxy listening on unix:///var/run/weave.sock
INFO: 2015/09/30 11:12:50.117858 GET /v1.15/version
INFO: 2015/09/30 11:12:50.118814 GET /v1.15/_ping
INFO: 2015/09/30 11:12:50.125716 GET /v1.15/images/json?all=1
INFO: 2015/09/30 11:12:50.268669 GET /v1.15/version
INFO: 2015/09/30 11:12:50.296761 GET /v1.15/_ping
INFO: 2015/09/30 11:12:50.311062 GET /v1.15/containers/json?all=1&filters=%7B%22status%22%3A%5B%5D%7D
INFO: 2015/09/30 11:12:50.314112 DELETE /v1.15/containers/c9e5e0774ee4b7790e5ba39f8a3b1bd85c3106d49145b2ba24b1e76cc105caa1?force=1
INFO: 2015/09/30 11:12:50.826403 POST /v1.15/containers/create?hostconfig=%7B%22Binds%22%3A%5B%22%2Fvar%2Frun%2Fweave.sock%3A%2Fvar%2Frun%2Fdocker.sock%22%2C%22%2Fvar%2Flog%2Fecs%3A%2Flog%22%2C%22%2Fvar%2Flib%2Fecs%2Fdata%3A%2Fdata%22%2C%22%2Fetc%2Fecs%3A%2Fetc%2Fecs%22%2C%22%2Fvar%2Fcache%2Fecs%3A%2Fvar%2Fcache%2Fecs%22%2C%22%2Fcgroup%3A%2Fcgroup%3Aro%22%2C%22%2Fvar%2Frun%2Fdocker%2Fexecdriver%3A%2Fvar%2Flib%2Fdocker%2Fexecdriver%3Aro%22%5D%2C%22PortBindings%22%3A%7B%2251678%2Ftcp%22%3A%5B%7B%22HostIP%22%3A%22127.0.0.1%22%2C%22HostPort%22%3A%2251678%22%7D%5D%7D%2C%22RestartPolicy%22%3A%7B%7D%7D&name=ecs-agent
INFO: 2015/09/30 11:12:50.826642 Creating container with WEAVE_CIDR ""
INFO: 2015/09/30 11:12:50.990839 POST /v1.15/containers/a76874480e48949c309c0b644f54e832f268a5115cbd8c52f64d9b1120b2ee94/start
2015/09/30 11:12:50 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
goroutine 34 [running]:
net/http.func·011()
	/usr/local/go/src/net/http/server.go:1130 +0xbb
github.com/weaveworks/weave/proxy.(*Proxy).addWeaveWaitVolume(0xc20805cb00, 0x0)
	/home/go/src/github.com/weaveworks/weave/proxy/proxy.go:392 +0x4d9
github.com/weaveworks/weave/proxy.(*startContainerInterceptor).InterceptRequest(0xc2080367c0, 0xc2092fa410, 0x0, 0x0)
	/home/go/src/github.com/weaveworks/weave/proxy/start_container_interceptor.go:31 +0x2f5
github.com/weaveworks/weave/proxy.(*Proxy).Intercept(0xc20805cb00, 0x7ff501ad2730, 0xc2080367c0, 0x7ff501ad2068, 0xc2090fbea0, 0xc2092fa410)
	/home/go/src/github.com/weaveworks/weave/proxy/proxy_intercept.go:18 +0x4c
github.com/weaveworks/weave/proxy.(*Proxy).ServeHTTP(0xc20805cb00, 0x7ff501ad2068, 0xc2090fbea0, 0xc2092fa410)
	/home/go/src/github.com/weaveworks/weave/proxy/proxy.go:167 +0x259
net/http.serverHandler.ServeHTTP(0xc20800d4a0, 0x7ff501ad2068, 0xc2090fbea0, 0xc2092fa410)
	/usr/local/go/src/net/http/server.go:1703 +0x19a
net/http.(*conn).serve(0xc2090fbe00)
	/usr/local/go/src/net/http/server.go:1204 +0xb57
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1751 +0x35e
```

### PR DESCRIPTION
**Pull request by 2opremio**
Fixes #1481

### CODE REVIEW

**Comment by 2opremio** [comment]
@paulbellamy Please review

**Comment by paulbellamy** [comment]
Could you just remove the `&` on the unmarshal call? would that be simpler?

**Comment by paulbellamy** [comment]
Also, what's the `-o /dev/null -w "%{http_code}\n"` to curl?

**Comment by paulbellamy** [comment]
Also, does curl fail on response codes >= 400?

File: test/config.sh [hunk 1]
```diff
@@ -106,7 +106,7 @@ docker_api_on() {
     data=$4
     shift 4
     [ -z "$DEBUG" ] || greyly echo "Docker (API) on $host:$DOCKER_PORT: $method $url" >&2
-    echo -n "$data" | curl -s -X "$method" -H Content-Type:application/json "http://$host:$DOCKER_PORT/v1.15$url" -d @-
+    echo -n "$data" | curl -s -o /dev/null -w "%{http_code}\n" -X "$method" -H Content-Type:application/json "http://$host:$DOCKER_PORT/v1.15$url" -d @-
```
Comment (rade) [inline test/config.sh]: I really would expect `docker_api_on` to return the json result of the API call, not the http status code.

**Comment by 2opremio** [comment]
> Could you just remove the `&` on the unmarshal call? would that be simpler?

Yep, that would be simpler, good point.

> Also, what's the `-o /dev/null -w "%{http_code}\n"` to curl?

It prints the status code of the curl request

File: test/config.sh (hunk 1 above)
Comment (2opremio) [inline test/config.sh]: Yep, I see your point, but I would like to distinguish when requests fail. Docker doesn't report errors in the response body (at least for [start](https://docs.docker.com/reference/api/docker_remote_api_v1.15/#start-a-container))

Comment (rade) [inline test/config.sh]: `curl -f`?

Comment (2opremio) [inline test/config.sh]: OK, some responses have a body. How about a specialization? `docker_api_status_on`

Or even check for the returned status code in `docker_api_on`  and return whether it's 204?

Comment (2opremio) [inline test/config.sh]: > `curl -f`?

Yep, `curl -f` would implicitly do the same as the second suggestion I made. I'll do that.

**Comment by 2opremio** [comment]
@paulbellamy @rade PTAL

File: proxy/start_container_interceptor.go [hunk 2]
```diff
@@ -35,7 +35,7 @@ func (i *startContainerInterceptor) InterceptRequest(r *http.Request) error {
 		}
 	}
 
-	if err := marshalRequestBody(r, hostConfig); err != nil {
+	if err := marshalRequestBody(r, *hostConfig); err != nil {
```
Comment (paulbellamy) [inline proxy/start_container_interceptor.go]: Do we need this? Can't we marshal the pointer to a hostconfig the same?

Comment (2opremio) [inline proxy/start_container_interceptor.go]: Uhm, it is not completely equivalent (note that the reason for this PR was precisely that unmarshalling JSON leads to a different result for pointers and structs).

Probably in this case it doesn't make a difference (we guarantee that the pointer is not null), but I did it to make it explicit.

Comment (rade) [inline proxy/start_container_interceptor.go]: > we guarantee that the pointer is not null

if it was null this would explode, wouldn't it?

Comment (2opremio) [inline proxy/start_container_interceptor.go]: Yes

**Pull request merged**